include/domain_bridge/service_bridge_impl.inc

```diff
@@ -0,0 +1,202 @@
+    // send_response(*request_header, *response);
+    // ***
+    // The hack: forget about sending a response immediately.
+    callback_(this->shared_from_this(), request_header, typed_request);
+  }
+private:
+  // any_callback_ is private in service, we will use our own callback implementation.
+  CallbackT callback_;
+};
+
+/// Create a service with a given type.
+/// \internal
+template<typename ServiceT>
+typename std::shared_ptr<HackedService<ServiceT>>
+create_hacked_service(
+  std::shared_ptr<rclcpp::Node> node,
+  const std::string & service_name,
+  typename HackedService<ServiceT>::CallbackT callback,
+  rclcpp::CallbackGroup::SharedPtr group)
+{
+  rcl_service_options_t service_options = rcl_service_get_default_options();
+  auto serv = std::make_shared<HackedService<ServiceT>>(
+    node->get_node_base_interface()->get_shared_rcl_node_handle(),
+    service_name, std::move(callback), service_options);
+  auto serv_base_ptr = std::static_pointer_cast<rclcpp::ServiceBase>(serv);
+  node->get_node_services_interface()->add_service(serv_base_ptr, group);
+  return serv;
+}
+
+rclcpp::Node::SharedPtr
+get_node_for_domain(DomainBridgeImpl & impl, std::size_t domain_id);
+
+const std::string &
+get_node_name(const DomainBridgeImpl & impl);
+}  // namespace detail
+
+/// Bridge a service from one domain to another.
+/**
+ * \param service_bridge: Struct containing info about the service to bridge.
+ * \param options: Options for bridging the topic.
+ */
+template<typename ServiceT>
+void
+DomainBridge::bridge_service(
+  const TopicBridge & service_bridge,
+  const TopicBridgeOptions & options)
+{
+  const auto & node_name = detail::get_node_name(*impl_);
+  // Validate service name
+  const std::string & service_name = rclcpp::expand_topic_or_service_name(
+    service_bridge.topic_name, node_name, "/", true);
+
+  // If a remap name is provided, then validate it
+  // otherwise "remap" to the same name
+  std::string service_remapped = service_name;
+  if (!options.remap_name().empty()) {
+    service_remapped = rclcpp::expand_topic_or_service_name(
+      options.remap_name(), node_name, "/", true);
+  }
+
+  const std::size_t & from_domain_id = service_bridge.from_domain_id;
+  const std::size_t & to_domain_id = service_bridge.to_domain_id;
+
+  // Check if already bridged
+  if (service_bridge_impl_->bridged_services_.find(service_bridge) !=
+    service_bridge_impl_->bridged_services_.end())
+  {
+    std::cerr << "Service '" << service_name << "'" << " already bridged from domain " <<
+      std::to_string(from_domain_id) << " to domain " << std::to_string(to_domain_id) <<
+      ", ignoring" << std::endl;
+    return;
+  }
+
+  rclcpp::Node::SharedPtr from_domain_node = detail::get_node_for_domain(*impl_, from_domain_id);
+  rclcpp::Node::SharedPtr to_domain_node = detail::get_node_for_domain(*impl_, to_domain_id);
+
+  // Create a client for the 'from_domain'
+  auto client = from_domain_node->create_client<ServiceT>(
```

Comment:
Isn't this backwards? I would expect requests to be received by the bridge on the "from_domain" and then forwarded to the "to_domain". Maybe elaborating in the design doc (optionally with a diagram) would help clarify.

Reply:
I think it depends of how you think about it:

- You're bridging a service server from domain A to domain B.
- You're bridging requests from domain A to domain B.

Maybe we can be more specific:

- original_server_domain_id
- new_server_domain_id

Those aren't great either, so I'm not sure.
@jacobperron feel free to suggest other naming alternatives.

Reply:
I see how this can be confusing. In my mind, I make the analogy between pub/sub and client/service so I had been thinking that we are bridging the *requests* from one domain into another.

However we decide to interpret a service bridge, we should make sure it is clearly documented in the API and in the design doc. I think a simple diagram of a service bridge to include in the design doc would be nice.

Reply:
Sounds good, I will add that to the docs.

I checked the `ros1_bridge`, and I think that it's using a similar terminology to the one here (https://github.com/ros2/ros1_bridge/blob/83390800c5ff205625428f148b514008932cfc61/include/ros1_bridge/factory.hpp#L340), so at least we're being consistent.

Reply:
https://github.com/ros2/domain_bridge/pull/26/commits/a1070609a9f9cfa7acb496e1a32fb38e24ed91c3, https://github.com/ros2/domain_bridge/pull/26/commits/f9fddf4ccae91b3a48bb42fe136058847a2c1eff and https://github.com/ros2/domain_bridge/pull/26/commits/40f1198e69749fdc24bed18ab227a613ee0d1639 added docs explaining how to bridge services.